src/d3a/models/market/two_sided.py

```diff
@@ -261,30 +261,53 @@ def match_recommendations(
-
-            if not market_offer:
-                raise OfferNotFoundException
-            if not market_bid:
-                raise BidNotFoundException
-
-            original_bid_rate = market_bid.original_bid_price / market_bid.energy
-            trade_bid_info = TradeBidOfferInfo(
-                original_bid_rate=original_bid_rate,
-                propagated_bid_rate=market_bid.price / market_bid.energy,
-                original_offer_rate=market_offer.original_offer_price / market_offer.energy,
-                propagated_offer_rate=market_offer.price / market_offer.energy,
-                trade_rate=original_bid_rate)
-
-            bid_trade, offer_trade = self.accept_bid_offer_pair(
-                market_bid, market_offer, recommended_pair["trade_rate"],
-                trade_bid_info, selected_energy
-            )
-            if offer_trade.residual is not None or bid_trade.residual is not None:
-                recommendations = self._replace_offers_bids_with_residual_in_recommendations_list(
-                    recommendations, offer_trade, bid_trade
+            clearing_rate = recommended_pair["trade_rate"]
+            market_offers = [
+                self.offers.get(offer["id"]) for offer in recommended_pair["offers"]]
+            market_bids = [self.bids.get(bid["id"]) for bid in recommended_pair["bids"]]
+
+            if not all(list(market_offers)):
+                # If not all received offers exist in the market, skip the current recommendation
+                continue
+            if not all(list(market_bids)):
+                # If not all received bids exist in the market, skip the current recommendation
+                continue
+
+            self.validate_bid_offer_match(
+                market_bids, market_offers,
+                clearing_rate, selected_energy)
+
+            market_offers = iter(market_offers)
+            market_bids = iter(market_bids)
+            market_offer = next(market_offers)
+            market_bid = next(market_bids)
+            while True:
+                original_bid_rate = market_bid.original_bid_price / market_bid.energy
+                trade_bid_info = TradeBidOfferInfo(
+                    original_bid_rate=original_bid_rate,
+                    propagated_bid_rate=market_bid.price / market_bid.energy,
+                    original_offer_rate=market_offer.original_offer_price / market_offer.energy,
+                    propagated_offer_rate=market_offer.price / market_offer.energy,
+                    trade_rate=original_bid_rate)
+
+                bid_trade, offer_trade = self.accept_bid_offer_pair(
+                    market_bid, market_offer, clearing_rate,
+                    trade_bid_info, selected_energy)
+                if offer_trade.residual:
+                    market_offer = offer_trade.residual
+                else:
+                    market_offer = next(market_offers, None)
```

Comment:
I dont understand why this is done here. Isn't the `market_offer` overwritten in Line 282 again in the next iteration?

Reply:
No, the one being reassigned later is the market_bid

Reply:
Sorry, this does not answer my question IMO. Can you please explain why this is done here more verbosely? Thanks!

Reply:
Sure
So this is how the algorithm works:

1. Receive a list of recommendations, each recommendation can have multiple bids against multiple offers
2. While this list is not empty:
3. pop the first recommendation `recommended_pair = recommendations.pop()`
4. Get the market bids/offers version of the ones in the recommendation
5. If any bid or offer id doesn't exist in the market, continue | move on to the next recommendation
6. validate_bid_offer_match is called to validate that these can be matched
7. Make the bids/offers of the current recommendation as iterable
8. Pick the first bid and the first offer using next()
8. In a while True loop, we match the formers together
9. If there is a bid residual, make the market_bid = bid_trade.residual, the same goes for the offer (Otherwise we will pick the next item in the iterable)
10. We will always be matching the market_bid and market_offer until one of two iterables are fully consumed, this will break the while True loop and we'll move on to the next recommendation

Reply:
Darn! I overlooked `while True` loop. I thought this is all in the `while recommendations` loop. Ok, Understood. Thanks for the explanation!